### PR TITLE
Added go-fuzz for parser

### PR DESCRIPTION
**Pull request by tie**
It’d be nice to integrate continuous fuzzing too, but I’m not aware of any easy to set up (and free) services.

See also #252

### CODE REVIEW

**Comment by asdine** [comment]
Could we run the fuzzer on a Github Action when a branch is merged on master for example?

**Comment by tie** [comment]
Yes, we can do that. I think it’s better to run a scheduled nightly fuzz test.

I’ll write a workflow that stores corpus in cache, but we’d probably need a separate repo or file store for `corpus` data if we don’t want to lose it.

**Comment by asdine** [comment]
@tie Is this something that must be generated once and stored somewhere? If so, how about storing it in the Genji repo directly?

**Comment by tie** [comment]
For each `Fuzz` function we need [corpus data](https://github.com/dvyukov/go-fuzz-corpus/tree/e6d670e/parser/corpus). Ideally, we should also update it since running go-fuzz may find new inputs that increase coverage (we can push from GitHub Actions workflow).

See https://github.com/google/fuzzing/blob/master/tutorial/libFuzzerTutorial.md#continuous-fuzzing

**Comment by asdine** [comment]
@tie I have created a repo for storing the corpus data: https://github.com/genjidb/go-fuzz-corpus

File: .github/workflows/test.yaml [hunk 1]
```diff
@@ -0,0 +1,57 @@
+name: Unit Tests
```
Comment (asdine) [inline .github/workflows/test.yaml:1]: Thanks! Let's remove the Travis config then

File: fuzz/fuzz.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+	t.Helper()
+
+	err := filepath.Walk("testdata", func(path string, info os.FileInfo, err error) error {
+		t.Helper()
+
+		if err != nil {
+			return err
+		}
+
+		if info.IsDir() && path == "suppressions" {
```
Comment (asdine) [inline fuzz/fuzz.go:22]: Could you add a comment here to explain why we are skipping this directory?

File: .github/workflows/test.yaml (hunk 1 above)
Comment (tie) [reply .github/workflows/test.yaml:1]: Note that `actions/setup-go` currently does not support gotip, though it should be easier to add now that they have the necessary infrastructure in [actions/go-versions](https://github.com/actions/go-versions).

File: fuzz/fuzz.go (hunk 2 above)
Comment (tie) [reply fuzz/fuzz.go]: I’ve updated walkFn so that it only visits regular files in corpus and crashers directories.

**Comment by asdine** [comment]
@tie Is it ready for review?

**Comment by tie** [comment]
@asdine yes!

There would be another PR to `go-fuzz-corpus` repo with actions workflow to update the corpus on nightly schedule but the changes for main repo are ready. Here are the latest workflow runs on the PR source branch.
- [Fuzz Tests](https://github.com/tie/genji/runs/1253839914?check_suite_focus=true)
- [Unit Tests](https://github.com/tie/genji/runs/1253840265?check_suite_focus=true)
  Note that the (1.15.x, windows-latest, genji/engine/badgerengine) combo fails with `-race` flag. Should we (temporarily) exclude it from the matrix?

File: .github/workflows/test.yaml (hunk 1 above)
Comment (tie) [reply .github/workflows/test.yaml]: I’ve added codecov upload step in GitHub Actions, but I think we should keep Travis for `gotip` tests for now.

**Comment by tie** [comment]
I’ve removed `go test` workflows since they are still work-in-progress. I’ll submit them as a new PR, and will also add `tinygo` test runner (currently we only run tests with `tinygo` tag in Travis config).

**Review by asdine — APPROVED** [review]
Thanks @tie ! That's great job! Let's iterate on this for tinygo and race on subsequent PRs 👍🏼

**Comment by tie** [comment]
Neat, it’s working. Found another bug too!

**Pull request merged**